crates/ruff_server/src/trace.rs

```diff
@@ -0,0 +1,127 @@
+use lsp_types::notification::Notification;
+use serde::Deserialize;
+use std::sync::OnceLock;
+use tracing::{level_filters::LevelFilter, Event};
+use tracing_subscriber::layer::SubscriberExt;
+
+use crate::server::ClientSender;
+
+static LOGGING_SENDER: OnceLock<ClientSender> = OnceLock::new();
+
+pub(crate) fn init_tracing(sender: &ClientSender, filter: LogLevel) {
+    LOGGING_SENDER
+        .set(sender.clone())
+        .expect("logging sender should only be initialized once");
+
+    let subscriber = tracing_subscriber::Registry::default().with(LogLayer { filter });
+
+    tracing::subscriber::set_global_default(subscriber)
+        .expect("should be able to set global tracing subscriberr");
+}
```

Comment:
I wonder what will happen now where `ClientSender` is a weakref. I suspect that we won't see the `Main loop stopped` message anymore because the logging sender is shut down by then. 

I think we should unset our tracing subscriber before dropping `Connection` and restore the old subscriber to ensure these messages don't get lost.

Reply:
> I think we should unset our tracing subscriber before dropping `Connection` and restore the old subscriber to ensure these messages don't get lost.

Good idea!